core/src/databases/table_schema.rs

```diff
@@ -262,6 +265,85 @@ impl TableSchema {
                 .collect::<Result<Vec<_>>>(),
         }
     }
+
+    pub fn merge(&self, other: &Self) -> Result<Self, anyhow::Error> {
+        // Index our and other columns by name.
+        let our_columns_set = self.0.iter().map(|c| &c.name).collect::<HashSet<_>>();
+        let other_columns_map = other
+            .0
+            .iter()
+            .map(|c| (&c.name, c))
+            .collect::<HashMap<_, _>>();
+
+        let merged_schema = self
+            .0
+            .iter()
+            // Iterate over all of our columns.
+            .map(|our_column| {
+                let mut our_column = our_column.clone();
+
+                // If the other schema has a column with the same name, merge it into our column.
+                if let Some(other_column) = other_columns_map.get(&our_column.name) {
+                    // If types are different, we need to merge them.
+                    if our_column.value_type != other_column.value_type {
+                        use TableSchemaFieldType::*;
+                        our_column.value_type =
+                            match (&our_column.value_type, &other_column.value_type) {
+                                // Ints and Floats can be merged into Floats.
+                                // Other types are widened to Text.
+                                (Int, Float) | (Float, Int) => TableSchemaFieldType::Float,
+                                _ => TableSchemaFieldType::Text,
+                            };
+                    }
+
+                    // Concatenate the unique possible values from both columns.
+                    our_column.possible_values = match our_column
+                        .possible_values
+                        .as_ref()
+                        .unwrap_or(&vec![])
+                        .iter()
+                        .chain(other_column.possible_values.as_ref().unwrap_or(&vec![]))
+                        .map(|v| v.to_string())
+                        .unique()
+                        .enumerate()
+                        .map(|(i, v)| {
+                            // If the total number of possible values is too large, or if any of the values are
+                            // too long, then we give up on possible values.
+                            // If there are no possible values, then we set it to None.
+                            if v.len() > POSSIBLE_VALUES_MAX_LEN || i >= POSSIBLE_VALUES_MAX_COUNT {
+                                None
+                            } else {
+                                Some(v)
+                            }
+                        })
+                        .collect::<Option<Vec<String>>>()
```

Comment:
this is the best rust feature i love it so much

Reply:
If my understanding is correct, as soon as there is one `None` in the `map` above, iteration is stopped because the `collect` can already get its final value (`None`)

Reply:
Never used it with Option<> did you double check the behavior here?